### ISSUE TITLE
vispy Canvas not redrawing after mouse click in MacOs (super old High Sierra 10.13.6)

### ISSUE THREAD

**Issue by olinickalls** [vispy/vispy#2309]
Hi-
 I really want to get this working on MacOs- it is running beautifully well on my windows PCs (Python 3.9.6).  Bear in mind this is an old late 2011 macbook pro, running High Sierra 10.13.6 - this is as far as I can update this old hardware...

Running both of these official vispy examples under MacOs Python 3.10.2 causes odd behavior:
  [isocurve_for_trisurface_qt](https://vispy.org/gallery/scene/isocurve_for_trisurface_qt.html)
  [displaying an image](https://vispy.org/gallery/scene/image.html)

The image and plot both display normally until you click on the canvas.  The image largely disappears but leaves some odd discoloured pixels behind.
The isocurve example canvas seems to disappear until you do something to force a change in window size when it seems to re-draw and display the canvas properly again.  This only seems to happen for clicking (left, middle or right buttons) - scrolling the wheel does not trigger this behaviour.

Image example - Before:
![image1](https://user-images.githubusercontent.com/47024733/157064657-53b3fe94-8527-4e86-8e9f-3db97004a654.jpeg)

After:
![image2](https://user-images.githubusercontent.com/47024733/157064685-5b3e758e-ee3c-40d4-afe5-6840c8ec7556.jpeg)
Note the slightly pink pixels on the left side - something is drawing.

Isocurve example- Before:
![isocurve1](https://user-images.githubusercontent.com/47024733/157064918-1c7c8a3f-87d4-4e19-8776-adb76bf84303.jpeg)

After:
![isocurve2](https://user-images.githubusercontent.com/47024733/157064931-36ac9e94-b8d3-4bb0-a700-8781f586245c.jpeg)
Note the lines at the bottom - I think they corresponded to the drawn curves.  I did a bit of a click-drag which is why the lines do not precisely correlate with the previous picture.

The canvas seems to _interact_ properly, it just seems as though the gui framework isn't redrawing the canvas properly.  A window resize triggers a proper re-draw and I see everything properly again.
I tried in separate code of mine which shows/hides a slider causing the canvas widget to shrink/grow without a main window resize- this causes a proper re-draw as well.

I have tried to force the widget update with `canvas.native.update()` and similar things with the main window to no avail.
`.hide()` followed by `.show()` works, but interrupts any drag operation.

I tried the isocurve example with pyside2, pyqt5 and the image example does not explicitly seem to use either framework (I'm not sure what it uses).
The same code works perfectly well in windows 10 under python 3.9.6.

I have seen some talk about different python versions causing different issues though I haven't seen this one in the issues list.

Any ideas how to diagnose, fix or mitigate this issue?

**Comment by djhoese** [comment]
That is a pretty old macbook, even for OpenGL standards, but let's see if we can nail this down. What version of vispy are you running? Can you try a Python 3.9 environment on the macbook? What version of PyQt5? Are you installing from pip or conda?

**Comment by olinickalls** [comment]
Thanks!
  Vispy 0.9.6
  PyQt5 5.15.6
  PySide2 5.15.2.1

PyQt6 and PySide6 don't have wheels for my ageing machine.

All installed from pip.

With python 3.9.10, there is exactly the same behaviour.
Interestingly my code (which disconnects the mouse_wheel event and connects to my own handler) seems ok when scrolling (it replaces a zoom fn with a prev/next slice fn) but the image demo just blanks.

I'll see if I can edit one of the demos to test this properly then post code.

**Comment by olinickalls** [comment]
Ok-
  Just adding this code into the image.py from above prevents the image blanking on mouse-wheel scrolling:

```
canvas.events.mouse_wheel.disconnect()

@canvas.events.mouse_wheel.connect
def on_mouse_wheel(event):
    print(f'Mouse wheel scroll captured: {event.delta}')
```
I wasn't really sure how useful this was, so I tried doing exactly the same as the native vispy code from the PanZoomTransform class which was originally capturing the mouse events:
[https://github.com/vispy/vispy/blob/main/vispy/visuals/transforms/interactive.py](https://github.com/vispy/vispy/blob/main/vispy/visuals/transforms/interactive.py) Line 90 and onwards:
```
    def on_mouse_wheel(self, event):
        """Mouse wheel handler
        Parameters
        ----------
        event : instance of Event
            The event.
        """
        self.zoom(np.exp(event.delta * (0.01, -0.01)), event.pos)
```

I just copied the last line and pushed it into my first function above, to become:
```
@canvas.events.mouse_wheel.connect
def on_mouse_wheel(event):
    print(f'Mouse wheel scroll captured: {event.delta}')
    view.camera.zoom(np.exp(event.delta * (0.01, -0.01)), event.pos)
```
This _should_ be doing exactly the same as the original native code. Except it doesn't.
The blanking issue with the wheel event no longer happens.

**Comment by djhoese** [comment]
You may be looking at the wrong piece of code (if I'm understanding things correctly). Here is the PanZoomCamera's mouse handler for mouse wheel:

https://github.com/vispy/vispy/blob/1df73c0d3683f806ed052c36a51ecb3b66382c86/vispy/scene/cameras/panzoom.py#L213-L216

Slightly different than what you are doing.

**Comment by olinickalls** [comment]
Ah- you are of course absolutely correct.

Using the correct code:

```
canvas.events.mouse_wheel.disconnect()

@canvas.events.mouse_wheel.connect
def on_mouse_wheel(event):
    print(f'Mouse wheel scroll captured: {event.delta}')

    center = view.camera._scene_transform.imap(event.pos)
    view.camera.zoom((1 + view.camera.zoom_factor)**(-event.delta[1] * 30), center)
    event.handled = True
```
The wheel behavior is perfect - same as the windows behaviour.


Apart from the `BaseCamera.viewbox_mouse_event(self, event)` at line 211 in panzoom.py, I don't see what is being done differently.  Even that call is just to a `pass` statement.

The weird thing is that this code - just re-connecting to the original handler - seems to fix the wheel issue:
```
canvas.events.mouse_wheel.disconnect()
canvas.events.mouse_wheel.connect(view.camera.viewbox_mouse_event)
```

I can only guess that there is a second connected 'slot' to the mouse_wheel 'signal' that is removed with the .disconnect() call.

**Comment by olinickalls** [comment]
With some hope I just tried:
```
canvas.events.mouse_press.disconnect()
canvas.events.mouse_press.connect(view.camera.viewbox_mouse_event)
```
Which has killed the button interaction - there is clearly more going on behind the scenes. :confused:

**Comment by djhoese** [comment]
If you are using a SceneCanvas then it also connects to https://github.com/vispy/vispy/blob/main/vispy/scene/canvas.py#L345. Besides handling the "picking" of Visuals I'm not sure it does much else. There should be some attributes on the event handler that can tell you if other things are connected.

**Comment by olinickalls** [comment]
Thanks for your help @djhoese - I think I have reached a working (if hacky) solution.

I looked into the surrounding code from your link above and quickly realised that I would need to develop a much greater understanding of the architecture of VisPy than I really am able to.  I dived a little way into the rabbit hole of VisPy event management, but attacking the problem that way round was just too confusing for me.

In the end I took the crude and somewhat pragmatic route of hijacking the signals again.

I disconnected the `mouse_press` signal and connected both `mouse_press` and `mouse_move` signals to my cut-down reimplementation of the `PanZoomCamera.viewbox_mouse_event()` method - essentially copy-pasted with updated namespace.

I realise this breaks much functionality- specifically when selecting items in the canvas.  As I add a connection without disconnecting `mouse_move` I think it preserves existing handling of those events, perhaps that is not important when the `mouse_press` handling has been bypassed.

For my purposes as a simple image stack viewer, I won't be interacting with the scene more than to pan/zoom/window/scroll.  I can live with the sacrifice.  I will also need to implement a mouse wheel handler to scroll through an image stack.

Final complete updated `image.py` code below.  Comments are generally preserved lines I cut or changed from the original `PanZoomCamera.viewbox_mouse_event()`.

```
import sys
from vispy import scene
from vispy import app
from vispy.io import load_data_file, read_png

import numpy as np

canvas = scene.SceneCanvas(keys='interactive')
canvas.size = 800, 600
canvas.show()

# Set up a viewbox to display the image with interactive pan/zoom
view = canvas.central_widget.add_view()

# Create the image
img_data = read_png(load_data_file('mona_lisa/mona_lisa_sm.png'))
interpolation = 'nearest'

image = scene.visuals.Image(img_data, interpolation=interpolation,
                            parent=view.scene, method='subdivide')

canvas.title = 'Spatial Filtering using %s Filter' % interpolation

# Set 2D camera (the camera will scale to the contents in the scene)
view.camera = scene.PanZoomCamera(aspect=1)
# flip y-axis to have correct aligment
view.camera.flip = (0, 1, 0)
view.camera.set_range()
view.camera.zoom(0.1, (250, 200))

# get interpolation functions from Image
names = image.interpolation_functions
names = sorted(names)
act = 17


# Implement key presses
@canvas.events.key_press.connect
def on_key_press(event):
    global act
    if event.key in ['Left', 'Right']:
        if event.key == 'Right':
            step = 1
        else:
            step = -1
        act = (act + step) % len(names)
        interpolation = names[act]
        image.interpolation = interpolation
        canvas.title = 'Spatial Filtering using %s Filter' % interpolation
        canvas.update()

def _my_process_mouse_event(event):  # original at https://github.com/vispy/vispy/blob/main/vispy/scene/canvas.py#L345
    """
    The SubScene received a mouse event; update transform
    accordingly.

    Parameters
    ----------
    event : instance of Event
        The event.
    """
    # if event.handled or not self.interactive:
    #     return

    # # Scrolling
    # BaseCamera.viewbox_mouse_event(self, event)

    # if event.type == 'mouse_wheel':
    #     center = self._scene_transform.imap(event.pos)
    #     self.zoom((1 + self.zoom_factor)**(-event.delta[1] * 30), center)
    #     event.handled = True

    if event.type == 'mouse_move':
        if event.press_event is None:
            return

        # modifiers = event.mouse_event.modifiers
        # p1 = event.mouse_event.press_event.pos
        # p2 = event.mouse_event.pos
        modifiers = event.modifiers
        p1 = event.press_event.pos
        p2 = event.pos

        if 1 in event.buttons and not modifiers:
            # Translate
            p1 = np.array(event.last_event.pos)[:2]
            p2 = np.array(event.pos)[:2]
            # p1s = self._transform.imap(p1)
            # p2s = self._transform.imap(p2)
            p1s = view.camera._transform.imap(p1)
            p2s = view.camera._transform.imap(p2)
            # self.pan(p1s - p2s)
            view.camera.pan(p1s - p2s)
            event.handled = True
        elif 2 in event.buttons and not modifiers:
            # Zoom
            p1c = np.array(event.last_event.pos)[:2]
            p2c = np.array(event.pos)[:2]
            # scale = ((1 + self.zoom_factor)**((p1c - p2c) *
            #                                     np.array([1, -1])))
            scale = ((1 + view.camera.zoom_factor)**((p1c - p2c) *
                                                np.array([1, -1])))
            # center = self._transform.imap(event.press_event.pos[:2])
            center = view.camera._transform.imap(event.press_event.pos[:2])
            # self.zoom(scale, center)
            view.camera.zoom(scale, center)
            event.handled = True
        else:
            event.handled = False
    elif event.type == 'mouse_press':
        # accept the event if it is button 1 or 2.
        # This is required in order to receive future events
        event.handled = event.button in [1, 2]
    else:
        event.handled = False

canvas.events.mouse_wheel.disconnect()
canvas.events.mouse_wheel.connect(view.camera.viewbox_mouse_event)

canvas.events.mouse_press.disconnect()
canvas.events.mouse_press.connect(_my_process_mouse_event)
canvas.events.mouse_move.connect(_my_process_mouse_event)



if __name__ == '__main__' and sys.flags.interactive == 0:
    app.run()
```

**Comment by djhoese** [comment]
@olinickalls Well I'm glad you got something working. Can we summarize this to make sure I'm understanding everything and maybe I have other suggestions depending on what happened. So...

1. You had a custom application that worked on a Windows box, but not an old macbook.
2. When you ran the originally mentioned examples including the basic image example **unchanged**, you got odd behavior when clicking on the Canvas. This was "fixed" by resizing the window. This was only a problem on the macbook.
3. You made a modified version of the image example that uses a copied version of the mouse handling from the PanZoomCamera for mouse press and mouse move events. Mouse wheel and mouse move are still both handled by the original camera object...but mouse move is also handled by your custom function.

I think we both got lost in the weeds trying to figure out what was going on rather than stepping back and noticing that the main issue is that the Canvas was only being partially drawn or drawn incorrectly. You had the issue with two examples that use different Visuals so it isn't the ImageVisual that is the problem.

This makes me think that some settings were being changed during the mouse handling that weren't being reset. This is very likely in the picking logic of the SceneCanvas. This logic essentially draws a Mesh inplace of your Visual to detect which Visual is clicked on. Related code:

SceneCanvas has a special render mode when picking:

https://github.com/vispy/vispy/blob/1df73c0d3683f806ed052c36a51ecb3b66382c86/vispy/scene/canvas.py#L478-L497

Which sets the `picking` property of all sub-visuals:

https://github.com/vispy/vispy/blob/1df73c0d3683f806ed052c36a51ecb3b66382c86/vispy/scene/visuals.py#L66-L74

Which stops the regular Visual from being drawn (unless `.interactive` is set on the individual Visual):

https://github.com/vispy/vispy/blob/1df73c0d3683f806ed052c36a51ecb3b66382c86/vispy/scene/visuals.py#L100-L103

But tells the ViewBox (a Widget subclass) to draw a Mesh:

https://github.com/vispy/vispy/blob/1df73c0d3683f806ed052c36a51ecb3b66382c86/vispy/scene/widgets/widget.py#L407-L416

I wonder if there is a Qt bug on your old laptop that is causing an exception during this picking step and it is getting swallowed up in the drawing steps and causes picking to never be properly turned off (until a resize).

**Comment by olinickalls** [comment]
Pretty much-
 I'd only add that mouse_wheel is being directly handled directly by the PanZoomCamera object, having been disconnected from the upstream event handling.

  I very quickly became lost in the picking code. All the 

  I have seen some difference between the same old mac and windows in another PySide2 app I wrote- I can't remember what it was now- at the time I just passed it off as me not understanding fundamental differences with the mac gui, and I just worked around it.  Maybe there's a bug I haven't put my finger on.

Thanks again for your help- I'm really enjoying using VisPy- I came from PyQtGraph but needed more performance displaying images with filtering and I'm very happy with the results!
